HDFS-16987. [BugFix] MarkBlockAsCorrupt should not mark a replica as corrupted if the DN has a newest replica

ZanderXu opened the pull request:
In our prod environment, we encountered an incident where HA failover caused some new corrupted blocks, causing some jobs to fail.

Traced down and found a bug in the processing of all pending DN messages when starting active services.
Suppose NN1 is Active and NN2 is Standby, Active works well and Standby is unstable

The steps to reproduce are as follows:

- Timing 1, client create a file, write some data and close it.

- Timing 2, client append this file, write some data and close it.

- Timing 3, Standby replayed the second closing edits of this file

- Timing 4, Standby processes the blockReceivedAndDeleted of the first create operation

- Timing 5, Standby processed the blockReceivedAndDeleted of the second append operation

- Timing 6, Admin switched the active namenode from NN1 to NN2

- Timing 7, client failed to append some data to this file.

```
org.apache.hadoop.ipc.RemoteException(java.io.IOException): append: lastBlock=blk_1073741825_1002 of src=/testCorruptedBlockAfterHAFailover is not sufficiently replicated yet.
    at org.apache.hadoop.hdfs.server.namenode.FSDirAppendOp.appendFile(FSDirAppendOp.java:138)
    at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.appendFile(FSNamesystem.java:2992)
    at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.append(NameNodeRpcServer.java:858)
    at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.append(ClientNamenodeProtocolServerSideTranslatorPB.java:527)
    at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java)
    at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:621)
    at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:589)
    at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:573)
    at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:1227)
    at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1221)
    at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1144)
    at java.security.AccessController.doPrivileged(Native Method)
    at javax.security.auth.Subject.doAs(Subject.java:422)
    at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1953)
    at org.apache.hadoop.ipc.Server$Handler.run(Server.java:3170) 
```

ZanderXu committed:
HDFS-16987. [BugFix] ProcessQueuedMessages should verify corrupted blocks

Hexiaoqiao commented:
@ZanderXu thanks for your report. Was the root cause collect wrong `toCorrupt` set here when process Reported block?

ZanderXu commented:
> @ZanderXu thanks for your report. Was the root cause collect wrong `toCorrupt` set here when process Reported block?

Standby doesn't mark corrupted block, just enqueue that block report into PendingDataNodeMessages. 
Active NameNode tries to process all  messages of the PendingDataNodeMessages and try to mark corrupted, but here it doesn't consider a case that this corrupted block may be invalid, which the datanode has reported a healthy replica, so the pending message is validate, can just ignore it.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 46s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  42m 30s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   1m 27s |  |  trunk passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  compile  |   1m 22s |  |  trunk passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  checkstyle  |   1m  5s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 29s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m  7s |  |  trunk passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   1m 34s |  |  trunk passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  spotbugs  |   3m 36s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  26m  0s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m 18s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 24s |  |  the patch passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javac  |   1m 24s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 25s |  |  the patch passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  javac  |   1m 25s |  |  the patch passed  |
| -1 :x: |  blanks  |   0m  0s | [/blanks-eol.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/1/artifact/out/blanks-eol.txt) |  The patch has 1 line(s) that end in blanks. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply  |
| -0 :warning: |  checkstyle  |   0m 58s | [/results-checkstyle-hadoop-hdfs-project_hadoop-hdfs.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/1/artifact/out/results-checkstyle-hadoop-hdfs-project_hadoop-hdfs.txt) |  hadoop-hdfs-project/hadoop-hdfs: The patch generated 1 new + 112 unchanged - 0 fixed = 113 total (was 112)  |
| +1 :green_heart: |  mvnsite  |   1m 40s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   1m  4s |  |  the patch passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   1m 39s |  |  the patch passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  spotbugs  |   4m  5s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  27m 54s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 222m 19s |  |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 41s |  |  The patch does not generate ASF License warnings.  |
|  |   | 342m 55s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.42 ServerAPI=1.42 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/5583 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux bab7c3dfb068 4.15.0-206-generic #217-Ubuntu SMP Fri Feb 3 19:10:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 33d73c0b6ae0b05e1fe8dfc8fce7734d7b49618b |
| Default Java | Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/1/testReport/ |
| Max. process+thread count | 2501 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/1/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

Hexiaoqiao commented:
Thanks for your comment. @ZanderXu 
> this corrupted block may be invalid

I think `block` you mentioned here is one replia actually, right? I am confused why corrupted replica may be invalid. if so, the `toCorrupt` set collected by `processReportedBlock` should be not reliable, then we should improve `processReportedBlock`. Is it related with Active-Standby failover or not? Sorry I don't review this code segment of trunk carefully.

ayushtkn commented:
@ZanderXu Some confusion here, Are you talking about like: The block got marked as a corrupt replica but namenode didn't delete to due to its recent transition into Active namenode which marked all the storages as stale & then delayed deletion of the blocks till the next BR?

Some extension of [HDFS-15200](https://issues.apache.org/jira/browse/HDFS-15200)?

ZanderXu commented:
@Hexiaoqiao @ayushtkn Master, thanks for your comments. I try to explain this problem cleanly.

First, we can reproduce this problem through the following steps:
Supposing there is a cluster with one Active NameNode, one Standby NameNode, and three datanodes (DN1, DN2 and DN3).

1. Client create a file with three replicas, write some data and close it. (supposing this file only has one block, blk_1024_1001)
2. Client try to append some data into this file and close it. (supposing this file only has one block, blk_1024_1002)
3. Client try to append some data into this file again and close it. (supposing this file only has one block, block_1024_1003)
4. Standby is unstable, and it replay all edits first, such as replay blk_1024_1001, blk_1024_1002, blk_1024_1003
5. Then Standby received some blockReceivedAndDeleted requests from Datanode and process them in order [ (DN1, blk_1024_1001),  (DN2, blk_1024_1001),  (DN3, blk_1024_1001),  (DN1, blk_1024_1002), (DN2, blk_1024_1002), (DN3, blk_1024_1002), (DN1, blk_1024_1003), (DN2, blk_1024_1003), (DN3, blk_1024_1003) ]
6. Standby NameNode will put the report message for blk_1024_1001 and blk_1024_1002 into PendingDataNodeMessage, because the GS of the stored block is 1003, 1001 and 1002 is less than 1003, so standby consider that these report messages may be corrupt, so just put them into PendingDataNodeMessage.
7. Right now, the block status in StandbyNameNode are as following:
PendingDataNodeMessage:  (DN1, blk_1024_1001), (DN2, blk_1024_1001), (DN3, blk_1024_1001), (DN1, blk_1024_1002), (DN2, blk_1024_1002), (DN3, blk_1024_1002)
BlockMap: (DN1, blk_1024_1003), (DN2, blk_1024_1003), (DN3, blk_1024_1003)
8. Right now, the block status in ActiveNameNode is normal. BlockMap: (DN1, blk_1024_1003), (DN2, blk_1024_1003), (DN3, blk_1024_1003)
9. HA failover, Active -> Standby, Standby -> Active. During starting Active service, namenode will process all messages of PendingDataNodeMessage. Because the GS 1001 and 1002 are less then 1003, namenode  marked these pending message as corrupted block and put these replicas into the corruptReplicas list.
10. Right now, the block status in  ActiveNameNode are as following:
CorruptReplicas:  (DN1, blk_1024_1001), (DN2, blk_1024_1001), (DN3, blk_1024_1001), (DN1, blk_1024_1002), (DN2, blk_1024_1002), (DN3, blk_1024_1002)
BlockMap: (DN1, blk_1024_1003), (DN2, blk_1024_1003), (DN3, blk_1024_1003)
11. Active NameNode will try to remove some invalid corrupted block while processing block report or blockReceived report if the datanode has reported one healthy replica.

```
// add block to the datanode
    AddBlockResult result = storageInfo.addBlock(storedBlock, reportedBlock);

    int curReplicaDelta;
    if (result == AddBlockResult.ADDED) {
      curReplicaDelta =
          (node.isDecommissioned() || node.isDecommissionInProgress()) ? 0 : 1;
      if (logEveryBlock) {
        blockLog.info("BLOCK* addStoredBlock: {} is added to {} (size={})",
            node, storedBlock, storedBlock.getNumBytes());
      }
    } else if (result == AddBlockResult.REPLACED) {
      curReplicaDelta = 0;
      blockLog.warn("BLOCK* addStoredBlock: block {} moved to storageType " +
          "{} on node {}", storedBlock, storageInfo.getStorageType(), node);
    } else {
      // if the same block is added again and the replica was corrupt
      // previously because of a wrong gen stamp, remove it from the
      // corrupt block list.
      corruptReplicas.removeFromCorruptReplicasMap(block, node,
          Reason.GENSTAMP_MISMATCH);
      curReplicaDelta = 0;
      blockLog.debug("BLOCK* addStoredBlock: Redundant addStoredBlock request"
              + " received for {} on node {} size {}", storedBlock, node,
          storedBlock.getNumBytes());
    }
```


I said the pending message is invalid or "the corrupted block" is invalid means that the datanode has report one healthy replica for this block, such as:blk_1024_1003, the pending messages with small GS are invalid, such as: blk_1024_1001, blk_1024_1002.

So NameNode can judge that if these pending message is valid according the status of the stored block.

ZanderXu commented:
> 11. Active NameNode will try to remove some invalid corrupted block while processing block report or blockReceived report if the datanode has reported one healthy replica.

NameNode can remove these invalid corrupted blocks from the corruptReplicas list until the corresponding DN do block report if there are no balance or mover for this block. This time may be long, such as 6 hours or 24 hours.

ayushtkn commented:
@ZanderXu When the NN marks it as corrupt. Does it gets added to postponeBlock? Why does ```dfs.namenode.corrupt.block.delete.immediately.enabled``` deletes it immediately?

ZanderXu commented:
> When the NN marks it as corrupt

`blockManager.processAllPendingDNMessages()` while starting active service.

>  Does it gets added to postponeBlock? 

No, this case has nothing to do with postponeBlock and stale storage.

ZanderXu commented:
`blockManager.processAllPendingDNMessages()` should check if the pended message is a redundant report request. If it is,   namenode shouldn't mark this replica as a corrupted replica. 

```
  for (BlockToMarkCorrupt b : toCorrupt) {
      if (!removeInvalidCorruptedBlocks) {
        markBlockAsCorrupt(b, storageInfo, node);
      } else {
        // Try to remove this corrupted replica directly if this storage has reported
        // a healthy replica. This logic is same with addStoredBlock.
        DatanodeStorageInfo otherStorage = b.getStored().findStorageInfo(node);
        boolean alreadyExist = otherStorage == storageInfo;
        if (alreadyExist) {
          blockLog.info("BLOCK* processAndHandleReportedBlock: Redundant report request " +
                  "received for {} on node {} with status {} and size {}.", b.getStored(), node,
              b.getCorrupted(), b.getCorrupted().getNumBytes());
        } else {
          markBlockAsCorrupt(b, storageInfo, node);
        }
      }
    }
```

hadoop-yetus commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 46s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  42m 24s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   1m 31s |  |  trunk passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  compile  |   1m 22s |  |  trunk passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  checkstyle  |   1m  3s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 30s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m  7s |  |  trunk passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   1m 33s |  |  trunk passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  spotbugs  |   3m 35s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  26m  1s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m 19s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 25s |  |  the patch passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javac  |   1m 25s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 15s |  |  the patch passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  javac  |   1m 15s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   0m 53s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 21s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 54s |  |  the patch passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   1m 29s |  |  the patch passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  spotbugs  |   3m 29s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  25m 52s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 233m 25s |  |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 53s |  |  The patch does not generate ASF License warnings.  |
|  |   | 350m 48s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.42 ServerAPI=1.42 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/5583 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux 8bded21a15e3 4.15.0-206-generic #217-Ubuntu SMP Fri Feb 3 19:10:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / f60f15f82e5f410220466b0243df54f0cb3b0d92 |
| Default Java | Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/2/testReport/ |
| Max. process+thread count | 1998 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/2/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

Hexiaoqiao commented:
Got it. Thanks for the additional information. IIUC, the root cause includes,
a. Standby postpone process request (BRD) from DataNode due to some reason, 
b. Active and Standby failover,
c. Replica's timestamp has upgraded (because append or recovery lease or other reason),
this three conditions have to intersect, right?

I agree ZanderXu's proposal that we should not put the replica to `corruptReplicas` set if there was another newest replica has reported already, or one stale replica should ignore directly.

ZanderXu commented:
> a. Standby postpone process request (BRD) from DataNode due to some reason,
> b. Active and Standby failover,
> c. Replica's timestamp has upgraded (because append or recovery lease or other reason),
> this three conditions have to intersect, right?

Yeah.

ayushtkn commented:
I think it required the replica with older genstamp should also be on the same datanode as with the replica with the newer genstamp.
Are you able to reproduce it when the 1001 replicas and 1002 replicas are on different datanodes?
Else this block deletes corrupt replica immediately
```
      // the block is over-replicated so invalidate the replicas immediately
      invalidateBlock(b, node, numberOfReplicas);
```
If you debug your test and go inside invalidateBlocks().
```
      // we already checked the number of replicas in the caller of this
      // function and know there are enough live replicas, so we can delete it.
      addToInvalidates(b.getCorrupted(), dn);
      removeStoredBlock(b.getStored(), node);
```
This ```    removeStoredBlock(b.getStored(), node);```  removes ``node`` from the stored block which even contained the 1002 genstamp.

Since all three 1001 & 1002 are on 3 same DN. For first 2
```
    boolean minReplicationSatisfied = hasMinStorage(b.getStored(),
        numUsableReplicas);

```

this stays satisfied, but since the previous 2 removed the actual storage which contained 1002 from the blockMap to get rid of 1001, so in the last iteration, this comes false, and hence the last replica isn't deleted.

So, I feel for this to trigger the 1001 & 1002 need to be on same datanode

ZanderXu commented:
Yeah, for this case, all the replicas with different GS are on the same datanodes, as follows:
| block | DN1 | DN2 | DN3 |
| --- | --- | --- | --- |
| blk_1024_1001 | dn1 | dn2 | dn3 |
| blk_1024_1002 | dn1 | dn2 | dn3 |

Standby NameNode stored blk_1024_1002 in blocksMap through replaying edits from active first. When processing these BRD rpcs from datanode, standby namenode will postpone the message with small GS. After processed all BRD requests, the block status in Standby are as:

- PendingDataNodeMessages: [blk_1024_1001, dn1], [blk_1024_1001, dn2], [blk_1024_dn3]
- BlocksMap: [blk_1024_1002, (dn1, dn2, dn3)]

Whiling starting Active service, the namenode marked (blk_1024_1001, dn1), (blk_1024_1001, dn2) and (blk_1024_1001, dn3) as corrupted blocks,  and removed the dn1 and dn2 from the storage list. Currently, the status of this block are as:

- CorruptReplicas: [blk_1024_1001, dn1], [blk_1024_1001, dn2], [blk_1024_1001, dn3]
- BlocksMap: [blk_1024_1002, **(dn3)**]

the `markBlockAsCorrupt` will remove dn1, dn2 from the storage list for the stored block (blk_1024_1002), because `corruptedDuringWrite` is true. And it will not remove dn3 from the storage list for the stored block (blk_1024_1002), because the `corruptedDuringWrite` is false (`minReplicationSatisfied` is false).

So for this case, maybe `markBlockAsCorrupt` should do somethings to avoid its wrong action (remove the expected dn1 and dn2 from the storage list).
```
    if (storageInfo != null) {
      storageInfo.addBlock(b.getStored(), b.getCorrupted());
    }
```
If this step returned `AddBlockResult.ALREADY_EXIST`, means the datanode already has one newest replica, so the `markBlockAsCorrupt` should return immediately. 

@ayushtkn How about directly modify some logic of `markBlockAsCorrupt` to fix this problem? what do you think?

ayushtkn commented:
I think yes, we should fix the ``markBlockAsCorrupt`` method itself while maintaining the regular flow. If a newer block exists on the same storage, We should pass a flag to both ``countNodes`` and ``invalidateBlock`` to make sure the newer storage doesn't get removed

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 46s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  42m 13s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   1m 29s |  |  trunk passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  compile  |   1m 21s |  |  trunk passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  checkstyle  |   1m  4s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 30s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m  8s |  |  trunk passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   1m 32s |  |  trunk passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  spotbugs  |   3m 39s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  26m 18s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m 20s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 24s |  |  the patch passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javac  |   1m 24s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 15s |  |  the patch passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  javac  |   1m 15s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   0m 54s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 21s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 54s |  |  the patch passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   1m 27s |  |  the patch passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  spotbugs  |   3m 29s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  26m  4s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  | 266m 54s | [/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/3/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt) |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 43s |  |  The patch does not generate ASF License warnings.  |
|  |   | 384m 23s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.namenode.TestAddOverReplicatedStripedBlocks |
|   | hadoop.hdfs.TestBlocksScheduledCounter |
|   | hadoop.hdfs.server.datanode.fsdataset.impl.TestFsDatasetImpl |
|   | hadoop.hdfs.TestLeaseRecovery2 |
|   | hadoop.hdfs.shortcircuit.TestShortCircuitLocalRead |
|   | hadoop.hdfs.TestAbandonBlock |
|   | hadoop.hdfs.server.blockmanagement.TestOverReplicatedBlocks |
|   | hadoop.hdfs.server.blockmanagement.TestBlocksWithNotEnoughRacks |
|   | hadoop.hdfs.TestReconstructStripedFileWithRandomECPolicy |
|   | hadoop.hdfs.TestReadStripedFileWithDecoding |
|   | hadoop.hdfs.server.blockmanagement.TestBlockManager |
|   | hadoop.hdfs.TestClientProtocolForPipelineRecovery |
|   | hadoop.hdfs.TestDisableConnCache |
|   | hadoop.hdfs.TestDFSStripedOutputStreamWithRandomECPolicy |
|   | hadoop.hdfs.TestDFSStripedOutputStreamUpdatePipeline |
|   | hadoop.hdfs.TestDeadNodeDetection |
|   | hadoop.hdfs.TestReconstructStripedFile |
|   | hadoop.hdfs.tools.TestDFSAdmin |
|   | hadoop.hdfs.TestDFSRemove |
|   | hadoop.hdfs.TestSmallBlock |
|   | hadoop.hdfs.server.datanode.TestNNHandlesBlockReportPerStorage |
|   | hadoop.hdfs.TestReplication |
|   | hadoop.hdfs.server.namenode.ha.TestObserverNode |
|   | hadoop.hdfs.server.datanode.TestReadOnlySharedStorage |
|   | hadoop.hdfs.server.namenode.metrics.TestNameNodeMetrics |
|   | hadoop.hdfs.server.blockmanagement.TestPendingReconstruction |
|   | hadoop.hdfs.server.namenode.TestAddStripedBlocks |
|   | hadoop.hdfs.server.namenode.TestProcessCorruptBlocks |
|   | hadoop.hdfs.TestDFSClientSocketSize |
|   | hadoop.hdfs.server.datanode.TestNNHandlesCombinedBlockReport |
|   | hadoop.hdfs.shortcircuit.TestShortCircuitCache |
|   | hadoop.hdfs.server.namenode.TestNameNodeMXBean |
|   | hadoop.hdfs.server.datanode.TestCorruptMetadataFile |
|   | hadoop.hdfs.server.datanode.TestDirectoryScanner |
|   | hadoop.hdfs.TestIsMethodSupported |
|   | hadoop.hdfs.server.namenode.TestListCorruptFileBlocks |
|   | hadoop.hdfs.crypto.TestHdfsCryptoStreams |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.42 ServerAPI=1.42 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/5583 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux 70c474b88fee 4.15.0-206-generic #217-Ubuntu SMP Fri Feb 3 19:10:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / e6325d8a57ac8a50e3f2e5ba3f4615e1012fefce |
| Default Java | Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/3/testReport/ |
| Max. process+thread count | 2223 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/3/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 47s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| -1 :x: |  mvninstall  |  11m 25s | [/branch-mvninstall-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/4/artifact/out/branch-mvninstall-root.txt) |  root in trunk failed.  |
| +1 :green_heart: |  compile  |   1m 39s |  |  trunk passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  compile  |   1m 15s |  |  trunk passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  checkstyle  |   1m  3s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 24s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m  5s |  |  trunk passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   1m 27s |  |  trunk passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  spotbugs  |   3m 35s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  28m 48s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m 27s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 29s |  |  the patch passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javac  |   1m 29s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 14s |  |  the patch passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  javac  |   1m 14s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   0m 52s | [/results-checkstyle-hadoop-hdfs-project_hadoop-hdfs.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/4/artifact/out/results-checkstyle-hadoop-hdfs-project_hadoop-hdfs.txt) |  hadoop-hdfs-project/hadoop-hdfs: The patch generated 1 new + 112 unchanged - 0 fixed = 113 total (was 112)  |
| +1 :green_heart: |  mvnsite  |   1m 30s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 55s |  |  the patch passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   1m 29s |  |  the patch passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  spotbugs  |   3m 39s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  26m 28s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  | 242m 13s | [/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/4/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt) |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 46s |  |  The patch does not generate ASF License warnings.  |
|  |   | 332m  0s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.namenode.TestFsck |
|   | hadoop.hdfs.server.blockmanagement.TestBlockManager |
|   | hadoop.hdfs.TestRollingUpgrade |
|   | hadoop.hdfs.TestClientProtocolForPipelineRecovery |
|   | hadoop.hdfs.server.datanode.TestDataNodeHotSwapVolumes |
|   | hadoop.hdfs.server.datanode.TestNNHandlesBlockReportPerStorage |
|   | hadoop.hdfs.server.namenode.ha.TestObserverNode |
|   | hadoop.hdfs.TestDecommission |
|   | hadoop.hdfs.server.blockmanagement.TestPendingReconstruction |
|   | hadoop.hdfs.TestDecommissionWithBackoffMonitor |
|   | hadoop.hdfs.server.datanode.TestNNHandlesCombinedBlockReport |
|   | hadoop.hdfs.TestDFSStripedOutputStreamWithFailure |
|   | hadoop.hdfs.server.datanode.TestDirectoryScanner |
|   | hadoop.hdfs.TestLeaseRecoveryStriped |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.42 ServerAPI=1.42 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/5583 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux 40bdeedf9472 4.15.0-206-generic #217-Ubuntu SMP Fri Feb 3 19:10:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 2f7fa4b1d5894a54bf9524a1c2126f6aae6869a0 |
| Default Java | Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/4/testReport/ |
| Max. process+thread count | 2073 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/4/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

ZanderXu commented:
@Hexiaoqiao @ayushtkn Master, after deep thinking, maybe we can only fix this problem when processAllPendingDNMessages, because namenode doesn't know whether this report is consistent with the actual replica storage information of the DataNode.

**Case1: This report with small GS is postponed report, which is different from the actual replica of the datanode.**
For example:

- The actual replica of DN is: blk_1024_1002
- The postponed report is: blk_1024_1001

For this case, namenode can ignore this postponed report and doesn't mark it as a corrupted replica. 

**Case2: This report with small GS is the newest report, which is same with the actual replica of the datanode.**
For example:

- The actual replica of DN is: blk_1024_1001
- The report is: blk_1024_1001
- The storages of this block in namenode already contains this DN

For this case, namenode shouldn't ignore this report, and it should mark this replica as a corrupted replica.  Manually modifying block storage files on DataNode may cause this problem.


At present,  namenode can only consider that each report is the newest report, and then modify the status of the block in the memory of namenode, because datanode reports the state  to NN through block report or blockReceiveAndDelete. 


If we modify the logic of `markBlockAsCorrupt`, namenode will can not mark the replica as a corrupted replica for case2.
If we modify the logic of `processAllPendingDNMessages`, the postponed message will be temporarily ignored for case 2, and active namenode will mark it as a corrupted replica in the next block report of corressponding DN.

Hexiaoqiao commented:
@ZanderXu thanks for your detailed comments. IIUC, the replica's GS is monotonically increasing at DataNode side, right? If that I think we could improve it rely on the basic rule at NameNode side,
a. ignore one replica with older GS at the same DataNode.
b. mark corrupt when one replica with same GS NameNode maintained but older GS compare to other replicas.
If this is true, improve `markBlockAsCorrupt` could fix this issue. Please correct me if I missing some information. Thanks.

ZanderXu commented:
> IIUC, the replica's GS is monotonically increasing at DataNode side, right?

Should we consider the case of manual destruction? such as, manually move the meta file to a small GS file.

ZanderXu commented:
And this problem only happens when the following conditions occur at the same time:

1. StandbyNameNode postpone some reports with small GS.
2. Transferring StandbyNameNode to ActiveNameNode. 

The `markBlockAsCorrupt` only be used by ActiveNameNode, and this method looks good to me, because Active can only only believe that every report is the newest report from datanode. And Active NameNode should modify state of block in memory according to these reports.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 53s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  | 242m 52s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   1m 19s |  |  trunk passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  compile  |   1m 10s |  |  trunk passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  checkstyle  |   1m  5s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 28s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   2m  9s |  |  trunk passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   1m 34s |  |  trunk passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  spotbugs  |   3m 27s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  25m 59s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m 10s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 14s |  |  the patch passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javac  |   1m 14s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  4s |  |  the patch passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  javac  |   1m  4s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   0m 51s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 11s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 53s |  |  the patch passed with JDK Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1  |
| +1 :green_heart: |  javadoc  |   1m 27s |  |  the patch passed with JDK Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09  |
| +1 :green_heart: |  spotbugs  |   3m 22s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  28m 33s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  | 224m 52s | [/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/5/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt) |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 42s |  |  The patch does not generate ASF License warnings.  |
|  |   | 546m  6s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.datanode.TestDirectoryScanner |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.42 ServerAPI=1.42 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/5583 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux 3c6e8bf45fef 4.15.0-206-generic #217-Ubuntu SMP Fri Feb 3 19:10:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 1dec31aea550107e40316ffab6f0f8210baa1456 |
| Default Java | Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.18+10-post-Ubuntu-0ubuntu120.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_362-8u362-ga-0ubuntu1~20.04.1-b09 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/5/testReport/ |
| Max. process+thread count | 2504 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5583/5/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

ZanderXu commented:
Modify some logic of `markBlockAsCorrupt` directly may can not solve the following abnormal case:

- Supposing there is a block, blk_1024_1002 with 3 replicas, and these replicas are stored in DN1, DN2 and DN3
- One day, some one changes the meta file of this block on DN2, such as delete the original meta file blk_1024_1002 and create an new meta file blk_1024_1001 with some random data
- The DN2 will report this replica to namenode after directory scanner or volume scanner found this inconsistent replica
- The nameNode will doesn't mark this replica as a corrupt replica because the storage list already contains DN2 if we modify `markBlockAsCorrupt` directly

So I think it looks better to modify `processAllPendingDNMessages` to solve this problem.

@Hexiaoqiao @ayushtkn Looking forward your good idea.

Hexiaoqiao commented:
Sorry for the late comments.

> manually move the meta file to a small GS file.

I don't think it is a good idea to process it manually.

> The nameNode will doesn't mark this replica as a corrupt replica because the storage list already contains DN2 if we modify markBlockAsCorrupt directly.

I didn't get it. Do you mean that it will not mark corrupt if improve the logic of `markBlockAsCorrupt`?  If that do we have any chance to fix and keep the logic when datanode report bad blocks? Thanks.
